src/base.mo

```diff
@@ -0,0 +1,424 @@
+      let bytes = Blob.toArray(key);
+
+      let (node_, last_, old_leaf, pos_) = find(nodes, bytes);
+
+      var last = last_;
+      var node = node_;
+
+      if (old_leaf == 0) {
+        let ?leaf = newLeaf(leaves, key) else return null;
+
+        setChild(node, last, leaf);
+        return ?(leaf >> 1);
+      };
+
+      let index = old_leaf >> 1;
+      let old_key = getKey(leaves, index);
+      if (key == old_key) {
+        return ?index;
+      };
+
+      let old_bytes = Blob.toArray(old_key);
+      var pos = pos_;
+      label l loop {
+        let ?add = newInternalNode(nodes) else {
+          setChild(node, last, old_leaf);
+          return null;
+        };
+        setChild(node, last, add);
+        node := add;
+
+        let (a, b) = (keyToIndex(bytes, pos), keyToIndex(old_bytes, pos));
+        pos +%= bitlength;
+        if (a == b) {
+          last := a;
+        } else {
+          setChild(node, b, old_leaf);
+          let ?leaf = newLeaf(leaves, key) else return null;
+          setChild(node, a, leaf);
+          return ?(leaf >> 1);
+        };
+      };
+      Debug.trap("Unreacheable");
+    };
+
+    public func lookup(key : Blob) : ?(Blob, Nat) {
+      assert key.size() == args.key_size;
+      let { leaves; nodes } = regions();
+
+      let bytes = Blob.toArray(key);
+
+      let (_, _, old_leaf, _) = find(nodes, bytes);
+      if (old_leaf == 0) return null;
+      let index = old_leaf >> 1;
+      return if (getKey(leaves, index) == key) ?(getValue(leaves, index), Nat64.toNat(index)) else null;
+    };
+
+    class Iterator(nodes : Region, forward : Bool) {
+      let stack = Array.init<(Nat64, Nat64)>(args.key_size * 8 / Nat16.toNat(bitlength), (0, 0));
+      var depth = 1;
+      stack[0] := if (forward) (0, 0) else (0, root_aridity_ - 1);
+
+      func next_step(i : Nat64) : Nat64 {
+        if (forward) {
+          i + 1;
+        } else {
+          if (i != 0) i - 1 else root_aridity_;
+        };
+      };
+
+      public func next() : ?Nat64 {
+        let leaf = label l : ?Nat64 loop {
+          let (node, i) = stack[depth - 1];
+          let max = if (depth > 1) aridity_ else root_aridity_;
+          if (i < max) {
+            let child = getChild(nodes, node, i);
+            if (child == 0) {
+              stack[depth - 1] := (node, next_step(i));
+              continue l;
+            };
+            if (child & 1 == 1) {
+              stack[depth - 1] := (node, next_step(i));
+              break l(?(child >> 1));
+            };
+            stack[depth] := (child, if (forward) 0 else aridity_ - 1);
+            depth += 1;
+          } else {
+            if (depth == 1) break l null;
+            depth -= 1;
+            let (prev_node, prev_i) = stack[depth - 1];
+            stack[depth - 1] := (prev_node, next_step(prev_i));
+          };
+        };
+        leaf;
+      };
+    };
+
+    func entries_(forward : Bool) : Iter.Iter<(Blob, Blob)> {
```

Comment:
If the argument is `dir : { #forward; #reverse }` then the public functions below will be easier to read.